mypy/subtypes.py

```diff
@@ -131,24 +143,33 @@ def visit_instance(self, left: Instance) -> bool:
         if isinstance(right, TupleType) and right.fallback.type.is_enum:
             return is_subtype(left, right.fallback)
         if isinstance(right, Instance):
-            # NOTO: left.type.mro may be None in quick mode if there
+            if (left, right) in right.type.cache:
+                return True
+            # NOTE: left.type.mro may be None in quick mode if there
             # was an error somewhere.
             if left.type.mro is not None:
                 for base in left.type.mro:
                     if base._promote and is_subtype(
                             base._promote, self.right, self.check_type_parameter,
                             ignore_pos_arg_names=self.ignore_pos_arg_names):
+                        right.type.cache.add((left, right))
                         return True
             rname = right.type.fullname()
-            if not left.type.has_base(rname) and rname != 'builtins.object':
-                return False
-
-            # Map left type to corresponding right instances.
-            t = map_instance_to_supertype(left, right.type)
-
-            return all(self.check_type_parameter(lefta, righta, tvar.variance)
-                       for lefta, righta, tvar in
-                       zip(t.args, right.args, right.type.defn.type_vars))
+            # Always try a nominal check if possible,
+            # there might be errors that a user wants to silence *once*.
+            if ((left.type.has_base(rname) or rname == 'builtins.object') and
```

Comment:
Speed optimization: If supertype is in MRO, but type variable check fails, then don't fall back to structural, return ``False`` immediately (basically just replace ``if`` -> ``elif`` below).